lib/manageiq/api/common/paginated_response.rb

```diff
@@ -0,0 +1,88 @@
+class PaginatedResponse
+  attr_reader :limit, :offset
+
+  def initialize(base_query:, request:, limit: nil, offset: nil)
+    @base_query = base_query
+    @request    = request
+    @limit      = (limit || 100).to_i
+    @offset     = (offset || 0).to_i
+  end
+
+  def response
+    {
+      "meta"  => metadata_hash,
+      "links" => links_hash,
+      "data"  => records
+    }
+  end
+
+  private
+
+  def metadata_hash
+    @metadata_hash ||= {"count" => count}
+  end
+
+  def links_hash
+    @links_hash ||= {
+      "first" => link_to_first,
+      "last"  => link_to_last,
+      "prev"  => link_to_prev,
+      "next"  => link_to_next,
+    }
+  end
+
+  def link_to_first
+    link_with_new_offset(0)
+  end
+
+  def link_to_last
+    link_with_new_offset(max_limit_multiplier.clamp(0, Float::INFINITY) * limit)
+  end
+
+  def link_to_prev
+    return if current_limit_multiplier == 0
+
+    link_with_new_offset((current_limit_multiplier - 1) * limit)
+  end
+
+  def link_to_next
+    next_limit_multiplier = current_limit_multiplier + 1
+    return if next_limit_multiplier > max_limit_multiplier
+
+    link_with_new_offset(next_limit_multiplier * limit)
+  end
+
+  def link_with_new_offset(offset)
+    request_uri.dup.tap do |u|
+      u.query = query_hash_merge("offset" => offset.to_s)
+    end.to_s
+  end
+
+  def query_hash_merge(new_hash)
+    parsed_query.merge(new_hash).collect { |k, v| "#{k}=#{v}" }.join("&")
```

Comment:
Might need URI escaping on the k and v (unless they are already escaped coming in )

Reply:
There's also a method that does this for you.. CGI.www_encoded_form or something like that (Also URI.query_string or something )

Reply:
`Hash#to_query` seemed simplest